### PR TITLE
JIT: expand qmarks earlier

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Move qmark expansion to happen at the end of the importation phases. That way we don't have to model the messy semantics of qmarks in analysis and optimization phases.

For now the downstream bits that dealt with qmarks are left intact. They will be removed later.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@jakobbotsch @EgorBo FYI.

Local runs show quite mixed diffs.

Looks like quite a few cases where we don't handle the expanded code as well as we could.  Cleaning this up might turn this into a reliable code size improvement.

Going to flight this in CI to see if there are any obvious correctness issues and to gauge TP impact.

**Comment by jakobbotsch** [comment]
Another source of regressions can be what we discussed in #76439... local assertion prop can be more precise when qmarks aren't expanded yet. Not sure how important this is, though.

**Comment by EgorBo** [comment]
Wonder what happened here:
```diff
 ;  V01 OutArgs      [V01    ] (  1,  1   )  struct (32) [rsp+00H]   do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
 ;* V02 tmp1         [V02    ] (  0,  0   )     ref  ->  zero-ref    class-hnd "Inline return value spill temp"
 ;  V03 tmp2         [V03    ] (  3,  6   )     int  ->  [rsp+20H]   do-not-enreg[X] addr-exposed ld-addr-op "Inlining Arg"
-;* V04 tmp3         [V04    ] (  0,  0   )     ref  ->  zero-ref    single-def "CASTCLASS eval op1"
-;* V05 tmp4         [V05    ] (  0,  0   )     ref  ->  zero-ref    class-hnd "spilling QMark2"
-;  V06 tmp5         [V06,T00] (  2,  2   )     ref  ->  rcx         class-hnd "Inline stloc first use temp"
+;* V04 tmp3         [V04,T03] (  0,  0   )     ref  ->  zero-ref    single-def "CASTCLASS eval op1"
+;  V05 tmp4         [V05,T00] (  2,  4   )     ref  ->  rcx         class-hnd "spilling QMark2"
+;  V06 tmp5         [V06,T01] (  3,  3   )     ref  ->  rcx         class-hnd single-def "Inline stloc first use temp"
+;  V07 cse0         [V07,T02] (  2,  2   )     ref  ->  rcx         "CSE - moderate"
 ;
 ; Lcl frame size = 40
 
@@ -31,6 +32,8 @@ G_M36762_IG03:        ; bbWeight=1, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref,
        jl       SHORT G_M36762_IG06
        mov      rcx, 0xD1FFAB1E
        ; gcrRegs +[rcx]
+       test     rcx, rcx
+       je       SHORT G_M36762_IG07
        lea      r8, [rsp+20H]
        mov      edx, 1
        xor      r9d, r9d
@@ -38,7 +41,7 @@ G_M36762_IG03:        ; bbWeight=1, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref,
        ; gcrRegs -[rcx] +[rax]
        ; gcr arg pop 0
        nop      
-						;; size=48 bbWeight=1 PerfScore 8.50
+						;; size=53 bbWeight=1 PerfScore 9.75
 G_M36762_IG04:        ; bbWeight=1, epilog, nogc, extend
        add      rsp, 40
        ret      
@@ -56,8 +59,16 @@ G_M36762_IG06:        ; bbWeight=0, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref
        ; gcr arg pop 0
        int3     
 						;; size=7 bbWeight=0 PerfScore 0.00
+G_M36762_IG07:        ; bbWeight=0, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref
+       mov      ecx, 57
+       mov      edx, 84
+       mov      rax, 0xD1FFAB1E      ; code for System.ThrowHelper:ThrowArgumentException(int,int)
+       call     [rax]System.ThrowHelper:ThrowArgumentException(int,int)
+       ; gcr arg pop 0
+       int3     
+						;; size=23 bbWeight=0 PerfScore 0.00
 
-; Total bytes of code 95, prolog size 4, PerfScore 23.50, instruction count 21, allocated bytes for code 95 (MethodHash=b60d7065) for method System.Tests.Perf_Array:ArrayCreate1D():System.Array:this
+; Total bytes of code 123, prolog size 4, PerfScore 27.55, instruction count 28, allocated bytes for code 123 (MethodHash=b60d7065) for method System.Tests.Perf_Array:ArrayCreate1D():System.Array:this
 ; ============================================================
```

**Comment by AndyAyersMS** [comment]
> Wonder what happened here:

Looks like we end up with a side effect in a block we should be able to jump thread, and so RBO is less effective.

**Comment by AndyAyersMS** [comment]
Latest [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=291889&view=ms.vss-build-web.run-extensions-tab). TP is looking good, but still some work to do on CQ issues.

File: src/coreclr/jit/promotionliveness.cpp [hunk 1]
```diff
@@ -528,31 +499,14 @@ void PromotionLiveness::FillInLiveness()
...
-            }
-
-            if (qmark == nullptr)
-            {
-                for (GenTree* cur = stmt->GetTreeListEnd(); cur != nullptr; cur = cur->gtPrev)
+                // Skip liveness updates/marking for defs; they may be conditionally executed.
+                if ((cur->gtFlags & GTF_VAR_DEF) == 0)
                 {
                     FillInLiveness(life, volatileVars, cur->AsLclVarCommon());
                 }
```
Comment (jakobbotsch) [inline src/coreclr/jit/promotionliveness.cpp:508]: This looks like it was updated to the wrong side of the if, this was the code for when there is a QMARK.

**Pull request closed without merging**